gpu/preconditioner/block_jacobi_kernels.cu

```diff
@@ -166,6 +166,151 @@ GKO_ENABLE_IMPLEMENTATION_SELECTION(select_apply, apply);
 using compiled_kernels = syn::compile_int_list<1, 13, 16, 32>;
 
 
+template <typename IndexType>
+__device__ __forceinline__ bool has_same_nonzero_pattern(
+    const IndexType *prev_row_ptr, const IndexType *curr_row_ptr,
+    const IndexType *next_row_ptr)
+{
+    if (next_row_ptr - curr_row_ptr != curr_row_ptr - prev_row_ptr) {
+        return false;
+    }
+    for (; curr_row_ptr < next_row_ptr; ++prev_row_ptr, ++curr_row_ptr) {
+        if (*curr_row_ptr != *prev_row_ptr) {
+            return false;
+        }
+    }
+    return true;
+}
+
+template <typename IndexType>
+__global__ void find_natural_blocks_kernel(
+    size_type num_rows, int32 max_block_size, const IndexType *row_ptrs,
+    const IndexType *col_idx, IndexType *block_ptrs, size_type *num_blocks_arr)
+{
+    block_ptrs[0] = 0;
+    if (num_rows == 0) {
+        return;
+    }
+    size_type num_blocks = 1;
+    int32 current_block_size = 1;
+    for (size_type i = 1; i < num_rows; ++i) {
+        const auto prev_row_ptr = col_idx + row_ptrs[i - 1];
+        const auto curr_row_ptr = col_idx + row_ptrs[i];
+        const auto next_row_ptr = col_idx + row_ptrs[i + 1];
+        if (current_block_size < max_block_size &&
+            has_same_nonzero_pattern(prev_row_ptr, curr_row_ptr,
+                                     next_row_ptr)) {
+            ++current_block_size;
+        } else {
+            block_ptrs[num_blocks] =
+                block_ptrs[num_blocks - 1] + current_block_size;
+            ++num_blocks;
+            current_block_size = 1;
+        }
+    }
+    block_ptrs[num_blocks] = block_ptrs[num_blocks - 1] + current_block_size;
+    num_blocks_arr[0] = num_blocks;
+}
+
+
+template <typename ValueType, typename IndexType>
+size_type find_natural_blocks(std::shared_ptr<const GpuExecutor> exec,
+                              const matrix::Csr<ValueType, IndexType> *mtx,
+                              int32 max_block_size, IndexType *block_ptrs)
+{
+    auto rows = mtx->get_num_rows();
+    auto cols = mtx->get_num_cols();
+    auto vals = mtx->get_const_values();
+    auto row_ptrs = mtx->get_const_row_ptrs();
+    auto col_idx = mtx->get_const_col_idxs();
+
+    Array<size_type> d_nums_array(exec, 1);
+    auto d_nums = d_nums_array.get_data();
+
+    Array<size_type> nums_array(exec->get_master(), 1);
+    auto nums = nums_array.get_data();
+
+    constexpr int64 blocksize3_1 = 1;
+    constexpr int64 blocksize3_2 = 1;
+    constexpr int64 blocksize3_3 = 1;
+
+    const int64 dimgrid3_1 = 1;
+    const int64 dimgrid3_2 = 1;
+    const int64 dimgrid3_3 = 1;
+
+    dim3 grid3(dimgrid3_1, dimgrid3_2, dimgrid3_3);
+    dim3 block3(blocksize3_1, blocksize3_2, blocksize3_3);
```

Comment:
Maybe we should just have:

```c++
find_natural_block_kernel<<<1, 1>>(....);
```
instead of having it spread out across this many lines? I think it's more obvious what we're doing that way.

You also have some strange type conversions going on here: [`dim3`](https://docs.nvidia.com/cuda/cuda-c-programming-guide/#dim3) is a triple of 32-bit unsigned integers, and you are unnecessarily converting to it from a signed 64-bit integer (after you've constructed a 64-bit signed integer from a 32-bit signed integer).

All that just to say you want to run something sequentially... seems like overkill to me.